docs/source/python/parquet.rst

```diff
@@ -604,3 +604,171 @@ One example is Azure Blob storage, which can be interfaced through the
 
     abfs = AzureBlobFileSystem(account_name="XXXX", account_key="XXXX", container_name="XXXX")
     table = pq.read_table("file.parquet", filesystem=abfs)
+
+Parquet Modular Encryption (Columnar Encryption)
+------------------------------------------------
+
+Columnar encryption is supported for Parquet files starting from
+Apache Arrow 4.0.0.
+
+Parquet uses the envelope encryption practice, where file parts are encrypted
+with "data encryption keys" (DEKs), and the DEKs are encrypted with "master
+encryption keys" (MEKs). The DEKs are randomly generated by Parquet for each
+encrypted file/column. The MEKs are generated, stored and managed in a Key
+Management Service (KMS) of user’s choice.
+
+Reading and writing encrypted parquet files involves passing file encryption
+and decryption properties to :class:`~pyarrow.parquet.ParquetWriter` and to
+:class:`~.ParquetFile`, respectively.
+
+Writing an encrypted parquet:
+
+.. code-block:: python
+
+   encryption_properties = crypto_factory.file_encryption_properties(
+                                    kms_connection_config, encryption_config)
+   with pq.ParquetWriter(filename, schema,
+                        encryption_properties=encryption_properties) as writer:
+      writer.write_table(table)
+
+Reading an encrypted parquet:
+
+.. code-block:: python
+
+   decryption_properties = crypto_factory.file_decryption_properties(
+                                                    kms_connection_config)
+   parquet_file = pq.ParquetFile(filename,
+                              decryption_properties=decryption_properties)
+
+
+In order to create the encryption and decryption properties, a ``CryptoFactory``
+should be created and initialized with KMS Client details, as described below.
+
+
+KMS Client
+~~~~~~~~~~
+
+The master encryption keys must be kept and managed in a production-grade KMS
+system, deployed in user's organization. Using Parquet encryption requires
+implementation of a client class for the KMS server.
+Any KmsClient implementation should implement the following informal interface:
+
+.. code-block:: python
+
+   class KmsClient:
+   def wrap_key(self, key_bytes, master_key_identifier):
+      """Wrap a key - encrypt it with the master key."""
+       raise NotImplementedError()
+
+   def unwrap_key(self, wrapped_key, master_key_identifier):
+      """Unwrap a key - decrypt it with the master key."""
+      raise NotImplementedError()
+
+An example KmsClient impelementation might look like the following:
+
+.. code-block:: python
+   class MyKmsClient(pq.KmsClient):
+      def __init__(self, kms_connection_configuration):
+         pq.KmsClient.__init__(self)
```

Comment:
The subclassing from pq.KmsClient is required, right? In that case, I would also add it to the minimal code snippet laying out the required methods above (or maybe combine the two code blocks into one)